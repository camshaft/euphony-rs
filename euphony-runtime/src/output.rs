use euphony_sc::project::Project;
use std::{io, sync::Arc};

pub type Handle = Arc<dyn Output>;

pub trait Output: 'static + Project + Send + Sync {
    fn finish(&self) -> io::Result<()>;
}
