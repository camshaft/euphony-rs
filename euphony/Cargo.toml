[package]
name = "euphony"
description = "music composition environment for programmers"
version = "0.1.1"
authors = ["Cameron Bytheway <bytheway.cameron@gmail.com>"]
edition = "2021"
license = "MIT"

[dependencies]
bach = "=0.0.6"
euphony-buffer = { version = "0.1", path = "../euphony-buffer", default-features = false }
euphony-command = { version = "0.1", path = "../euphony-command" }
euphony-macros = { version = "0.2", path = "../euphony-macros" }
euphony-samples = { version = "0.1", path = "../euphony-samples" }
euphony-units = { version = "0.1", path = "../euphony-units" }
futures = "0.3"
midly = { version = "0.5", default-features = false, features = ["std"] }
once_cell = "1"
pin-project = "1"
rand = { version = "0.8", default-features = false }

[dev-dependencies]
insta = "1"
euphony-node = { version = "0.1", path = "../euphony-node", features = ["reflect"] }
