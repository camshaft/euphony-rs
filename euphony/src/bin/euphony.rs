fn main() {
    euphony_cli::main()
}
