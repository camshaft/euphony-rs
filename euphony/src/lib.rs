pub use euphony_macros::*;
pub use euphony_units as units;

mod args;
pub mod ext;
pub mod output;
pub mod rand;
pub mod runtime;
pub mod section;
pub mod synth;

pub mod prelude {
    pub use crate::{
        ext::*,
        rand,
        runtime::{
            spawn, spawn_primary,
            time::{now, set_tempo, tempo},
        },
        section::section,
        units::{
            pitch::Interval,
            time::{Beat, Tempo},
        },
    };
    pub use futures::{FutureExt, StreamExt};
}

#[cfg(test)]
mod tests;
