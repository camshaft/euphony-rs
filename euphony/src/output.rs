pub use euphony_command::*;

pub(crate) use api::*;
pub use api::{set_file, set_stdout};
