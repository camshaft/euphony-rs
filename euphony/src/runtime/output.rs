use euphony_command::message::{self, Message};
use once_cell::sync::Lazy;
use std::{sync::Mutex, time::Duration};

pub trait Output: 'static + Send + Sync {
    fn emit(&mut self, message: Message);
}

struct Stdout;

impl Output for Stdout {
    fn emit(&mut self, message: Message) {
        println!("{}", message);
    }
}

static OUTPUT: Lazy<Mutex<Box<dyn Output>>> = Lazy::new(|| Mutex::new(Box::new(Stdout)));

pub fn set_output(output: Box<dyn Output>) {
    *OUTPUT.lock().unwrap() = output;
}

pub fn emit(message: Message) {
    OUTPUT.lock().unwrap().emit(message);
}

pub fn set_seed(seed: u64) {
    emit(Message::SetSeed { seed });
}

pub fn set_group_name(id: u64, name: &str) {
    emit(Message::SetGroupName { id, name });
}

pub fn advance(amount: Duration) {
    emit(Message::AdvanceTime { amount })
}

pub fn finish() {
    emit(Message::Finish);
}
