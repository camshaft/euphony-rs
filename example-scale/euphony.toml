[synths]
chiplead = "gh://sonic-pi-net/sonic-pi/etc/synthdefs/compiled/sonic-pi-chiplead.scsyndef"
bass = "gh://sonic-pi-net/sonic-pi/etc/synthdefs/compiled/sonic-pi-blade.scsyndef"

[samples]
bd = "clean://tidalcycles/sounds-tr808-fischer/bd8"
sd = "clean://tidalcycles/sounds-tr808-fischer/sd8"
hh = "clean://tidalcycles/sounds-tr808-fischer/ch8"
cy = "clean://tidalcycles/sounds-tr808-fischer/cy8"
ht = "clean://tidalcycles/sounds-tr808-fischer/ht8"
lt = "clean://tidalcycles/sounds-tr808-fischer/lt8"
mt = "clean://tidalcycles/sounds-tr808-fischer/mt8"
