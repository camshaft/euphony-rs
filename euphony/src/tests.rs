use crate as euphony;
use euphony::output;
use std::{
    future::Future,
    io::{self, Cursor},
    sync::{Arc, Mutex},
};

#[derive(Clone, Default)]
struct List(Arc<Mutex<Cursor<Vec<u8>>>>);

impl io::Write for List {
    fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
        self.0.lock().unwrap().write(buf)
    }

    fn flush(&mut self) -> io::Result<()> {
        Ok(())
    }
}

pub fn start<F>(name: &str, f: F)
where
    F: 'static + Future<Output = ()> + Send,
{
    let list = List::default();
    output::scope::set(Some(Box::new(list.clone())));

    euphony::runtime::Runtime::new(0).block_on(f);

    let mut result = core::mem::take(&mut *list.0.lock().unwrap());
    result.set_position(0);

    let mut dump = String::new();
    output::decode(&mut result, &mut dump).unwrap();

    insta::assert_display_snapshot!(name, dump);
}

use euphony::{prelude::*, synth::Processor};

static SINE: Processor = Processor {
    id: 100,
    name: "Sine",
    inputs: 3,
};
static SQUARE: Processor = Processor {
    id: 104,
    name: "Square",
    inputs: 3,
};

#[test]
fn tempo_test() {
    start("tempo_test", async {
        for i in 0..10 {
            set_tempo(Tempo(60 + i * 10, 1));
            let s = SINE.spawn();
            s.set(i % 3, now().as_f64());
            Beat(1, 2).delay().await;
        }
    })
}

#[test]
fn rand_test() {
    start("rand_test", async {
        // make sure we have a scope in the root
        let s = SINE.spawn();
        for _ in 0..5 {
            s.set(rand::gen_range(0..3), now().as_f64());
            Beat(1, 2).delay().await;
        }
        drop(s);

        // make sure spawned tasks have scopes
        async {
            let s = SINE.spawn();
            for _ in 0..5 {
                s.set(rand::gen_range(0..3), now().as_f64());
                Beat(1, 2).delay().await;
            }
        }
        .spawn()
        .await;

        // make sure we can seed tasks
        async {
            let s = SINE.spawn();
            for _ in 0..5 {
                s.set(rand::gen_range(0..3), now().as_f64());
                Beat(1, 2).delay().await;
            }
        }
        .seed(0)
        .spawn()
        .await;
    });
}

#[test]
fn scheduler_test() {
    start("scheduler_test", async {
        async {
            for i in 0..10 {
                let s = SINE.spawn();
                s.set(i % 3, now().as_f64());
                Beat(1, 2).delay().await;
            }
        }
        .spawn_primary();

        async {
            for i in 0..10 {
                let s = SQUARE.spawn();
                s.set(i % 3, now().as_f64());
                Beat(1, 4).delay().await;
            }
        }
        .spawn_primary();
    })
}
