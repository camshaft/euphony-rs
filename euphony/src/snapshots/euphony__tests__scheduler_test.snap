---
source: euphony/src/tests.rs
expression: dump
---
  SPN 0,0
  SET 0,0,0
  SPN 1,1
  SET 1,0,0
ADV 1024
  FIN 1
  SPN 2,1
  SET 2,1,0.25
ADV 1024
  FIN 0
  SPN 3,0
  SET 3,1,0.5
  FIN 2
  SPN 4,1
  SET 4,2,0.5
ADV 1024
  FIN 4
  SPN 5,1
  SET 5,0,0.75
ADV 1024
  FIN 3
  SPN 6,0
  SET 6,2,1
  FIN 5
  SPN 7,1
  SET 7,1,1
ADV 1024
  FIN 7
  SPN 8,1
  SET 8,2,1.25
ADV 1024
  FIN 6
  SPN 9,0
  SET 9,0,1.5
  FIN 8
  SPN 10,1
  SET 10,0,1.5
ADV 1024
  FIN 10
  SPN 11,1
  SET 11,1,1.75
ADV 1024
  FIN 9
  SPN 12,0
  SET 12,1,2
  FIN 11
  SPN 13,1
  SET 13,2,2
ADV 1024
  FIN 13
  SPN 14,1
  SET 14,0,2.25
ADV 1024
  FIN 12
  SPN 15,0
  SET 15,2,2.5
  FIN 14
ADV 2048
  FIN 15
  SPN 16,0
  SET 16,0,3
ADV 2048
  FIN 16
  SPN 17,0
  SET 17,1,3.5
ADV 2048
  FIN 17
  SPN 18,0
  SET 18,2,4
ADV 2048
  FIN 18
  SPN 19,0
  SET 19,0,4.5
ADV 2048
  FIN 19

